system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+        if (eval)
+        {
+            xmlChar *val = xmlXPathCastToString(eval);
+            xmlXPathFreeObject(eval);
+            if (val)
+            {
+                if (*val)
+                    ret = strToBool((const char *)val);
+                xmlFree(val);
+            }
+        }
+        return ret;
+    }
+
+    xmlNodePtr getSection(const char *name)
+    {
+        return getSectionNode(name, nullptr);
+    }
+    xmlNodePtr ensureSection(const char *name)
+    {
+        xmlNodePtr sect = getSection(name);
+        if (sect)
+            return sect;
+        return xmlNewChild(root, nullptr, (const xmlChar *) name, nullptr);
+    }
+    void removeSection(const char *name)
+    {
+        xmlXPathObjectPtr eval = xmlXPathEval((const xmlChar *) name, xpathCtx);
+        if (!eval)
+            return;
+        if (XPATH_NODESET==eval->type && !xmlXPathNodeSetIsEmpty(eval->nodesetval))
+        {
+            xmlNodeSetPtr ns = eval->nodesetval;
+            int count = xmlXPathNodeSetGetLength(ns);
+            for(int i=0; i<count; i++)
+            {
+                xmlNodePtr node = xmlXPathNodeSetItem(ns, i);
+                if (node)
+                {
+                    xmlUnlinkNode(node);
+                    xmlFreeNode(node);
+                    ns->nodeTab[i]=nullptr;
+                }
+            }
+        }
+        xmlXPathFreeObject(eval);
+    }
+    xmlNodePtr replaceSection(const char *name)
+    {
+        removeSection(name);
+        return xmlNewChild(root, nullptr, (const xmlChar *) name, nullptr);
+    }
+    virtual void setContent(const char *section, const char *xml) override
+    {
+        xmlNodePtr sect = replaceSection(section);
+        if (xml==nullptr) //means delete content
+            return;
+
+        xmlParserCtxtPtr parserCtx = xmlCreateDocParserCtxt((const unsigned char *)xml);
+        if (!parserCtx)
+            throw MakeStringException(-1, "CEsdlScriptContext:setXmlSection: Unable to init parse of %s XML content", section);
+//        parserCtx->myDoc = doc;
+        parserCtx->node = sect;
+        xmlParseDocument(parserCtx);
+        int wellFormed = parserCtx->wellFormed;
+        xmlFreeDoc(parserCtx->myDoc); //dummy document
+        xmlFreeParserCtxt(parserCtx);
+        if (!wellFormed)
+           throw MakeStringException(-1, "CEsdlScriptContext:setContent xml string: Unable to parse %s XML content", section);
+    }
+    virtual void setContent(const char *section, IPropertyTree *tree) override
+    {
+        xmlNodePtr sect = replaceSection(section);
+        if (tree==nullptr) //means delete content
+            return;
+        addChildFromPtree(sect, *tree, tree->queryName());
+    }
+
+    virtual void setAttribute(const char *section, const char *name, const char *value) override
+    {
+        xmlNodePtr sect = ensureSection(section);
+        xmlNewProp(sect, (const xmlChar *)name, (const xmlChar *)value);
```

Comment:
Not sure this is sufficient. Consider a use case where an attribute could be set to a default, then a subsequent operation intends to override the default. My observation is that the first operation succeeds and the subsequent operation is ignored. xmlSetProp will update a node, but may not (documentation isn't clear) insert.

On a tangent, it might also make sense to have operations that can remove an attribute (using xmlUnsetProp). The ability to remove a previously applied attribute might be the preferred approach to overriding.

Reply:
Agreed I'll change it to xmlSetProp.  I've looked at the code and that does create or update.

I'll also look at removal.

Reply:
Changing to xmlSetProp.  And changing this function to remove a prop if value is nullptr.  But leaving adding that capability to scripts for future consideration.